### PR TITLE
WIP Compile time checking for dereferencing nil with `ref` and changing not nil [ci skip]

### PR DESCRIPTION
**Pull request by alehander42**
Just opening the PR to make it easier to compare with upstream Nim: not ready

### CODE REVIEW

File: compiler/nilcheck.nim [hunk 1]
```diff
@@ -0,0 +1,667 @@
...
+  while not now.isNil:
+    if now.history.hasKey(name):
+      # h = h.concat(now.history[name])
+      return now.history[name]
+    now = now.previous
+  return @[]
+
+import macros
+
+macro aecho*(a: varargs[untyped]): untyped =
```
Comment (zah) [inline compiler/nilcheck.nim]: wouldn't a simpler definition work here:

```nim
template trace(args: varargs[untyped]) =
   when debugInfo:
      echo args
```

File: compiler/treetab.nim [hunk 2]
```diff
@@ -12,16 +12,18 @@
...
-  if n == nil: return
+proc hashTree*(n: PNode): Hash =
+  if n.isNil:
+    return
   result = ord(n.kind)
   case n.kind
   of nkEmpty, nkNilLit, nkType:
     discard
   of nkIdent:
+    echo "what", "ident"
```
Comment (zah) [inline compiler/treetab.nim]: Some debugging code left here.

File: compiler/treetab.nim [hunk 3]
```diff
@@ -12,16 +12,18 @@
...
     result = result !& n.ident.h
   of nkSym:
-    result = result !& n.sym.name.h
+    result = result !& n.sym.id
```
Comment (zah) [inline compiler/treetab.nim]: I recommended this change. 

What @alehander42 needed was a table being able to associate metadata to certain AST trees featuring symbols such as variables and fields. Treating symbols with the same name as equal seemed wrong and I expected that no code in the compiler is currently dependent on this behaviour.

File: tests/nilcheck/ref0.nim [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+  else:
+    A()
+
+proc f14(a: int): A not nil=
+  if a == 0:
+    return A()
+  else:
+    result = A()
+
+
```
Comment (zah) [inline tests/nilcheck/ref0.nim]: I assume you have these lines as "buffers" in order to be able to fix the line numbers in the error message above.

To solve this problem, I usually use the following process:

1) Write the code as properly formatted as possible.
2) Run the compiler to get the list of errors and it put in the spec above.
3) Re-run the compiler to produce the same list, but with corrected line numbers.
4) Copy-paste the new list of errors over the old one (the number of lines will stay the same and everything will be correct now)

File: compiler/nilcheck.nim (hunk 1 above)
Comment (alehander42) [reply compiler/nilcheck.nim]: yes, I agree, maybe I didn't know that varargs work in this situation

File: compiler/treetab.nim (hunk 2 above)
Comment (alehander42) [reply compiler/treetab.nim]: That should've been an assert: I wondered it it's possible to have a nkIdent there at all at this place in sempass

File: tests/nilcheck/ref0.nim (hunk 4 above)
Comment (alehander42) [reply tests/nilcheck/ref0.nim]: yeah, I started doing something like that in the later stages, it is a better approach(generally applicable to a lot of testing situations: run X which you know produces expected results and bootstrap tests from it)

**Comment by narimiran** [comment]
@alehander42 this has "work in progress" in the title: What's the status of this? Has there been any work or progress recently? Do you plan to continue to work on this?

**Comment by alehander42** [comment]
@narimiran not sure, the approach seems to work , it mostly needs rebasing and adding a pragma so it can be opt-in and more tests iirc, but I am not sure if its still relevant (maybe we need something that can also deal a bit with var/lent/other new pointers?)/ or not better done with CFG ? @Araq

**Comment by Araq** [comment]
Closing in favour of https://github.com/nim-lang/Nim/pull/12401

**Pull request closed without merging**